File: packages/python-host/src/ai37_agent_host/observability/trace_v1.py
"""Stable trace metadata envelope shared with the TypeScript host."""

from __future__ import annotations

import os
from typing import Any, Literal

TRACE_SCHEMA_VERSION = "trace.v1"
TraceKind = Literal["turn", "planner", "agent", "generation", "tool"]
TraceStatus = Literal["working", "input-required", "completed", "failed", "cancelled"]


def trace_metadata(
    trace_kind: TraceKind,
    *,
    turn_id: str,
    session_id: str,
    service: str | None = None,
    environment: str | None = None,
    **fields: Any,
) -> dict[str, Any]:
    """Build the canonical ``trace.v1`` metadata envelope."""
    return {
        **fields,
        "schemaVersion": TRACE_SCHEMA_VERSION,
        "traceKind": trace_kind,
        "service": service or "ai37-agent-host",
        "environment": environment
        or os.environ.get("LANGFUSE_TRACING_ENVIRONMENT")
        or os.environ.get("NODE_ENV")
        or "development",
        "turnId": turn_id,
        "sessionId": session_id,
    }

File: packages/ts-host/src/observability/trace-v1.ts
/** Stable metadata envelope shared by every AI37 host trace. */
export const TRACE_SCHEMA_VERSION = 'trace.v1' as const

export type TraceKind = 'turn' | 'planner' | 'agent' | 'generation' | 'tool'

export type TraceStatus =
  'working' | 'input-required' | 'completed' | 'failed' | 'cancelled'

export interface TraceMetadataV1 {
  schemaVersion: typeof TRACE_SCHEMA_VERSION
  traceKind: TraceKind
  service: string
  environment: string
  turnId: string
  sessionId: string
  taskId?: string
  runId?: string
  agentId?: string
  status?: TraceStatus
  intent?: string
  route?: string
  reasonCode?: string
  confidence?: number
  payloadMode?: 'inline' | 'inline-truncated' | 'external'
  [key: string]: unknown
}

export function traceMetadata(
  traceKind: TraceKind,
  fields: Omit<Partial<TraceMetadataV1>, 'schemaVersion' | 'traceKind'> &
    Pick<TraceMetadataV1, 'turnId' | 'sessionId'>,
): TraceMetadataV1 {
  return {
    ...fields,
    schemaVersion: TRACE_SCHEMA_VERSION,
    traceKind,
    service:
      typeof fields.service === 'string' && fields.service.length > 0
        ? fields.service
        : 'ai37-agent-host',
    environment:
      typeof fields.environment === 'string' && fields.environment.length > 0
        ? fields.environment
        : process.env.LANGFUSE_TRACING_ENVIRONMENT ||
          process.env.NODE_ENV ||
          'development',
  }
}
